### PR TITLE
Add type casting in `where` clause

### PR DESCRIPTION
**Pull request by andrykonchin**
Cast values in `where` clause to declared field types

Fix open issues https://github.com/Dynamoid/Dynamoid/issues/46 and https://github.com/Dynamoid/Dynamoid/issues/138

---

Looks like this PR breaks existed behaviour - search for field with custom type by `casted` value:

```
field :price, MoneyInstanceDump

Model.where(price: '5.0') 
```

It worked because we store value `5` in the storage and bypass type casting.
Now it shouldn't work because type casting requires something like this:

```
Model.where(price: MoneyInstanceDump.new('5.0')) 
```

### CODE REVIEW

File: lib/dynamoid/criteria/chain.rb [hunk 1]
```diff
@@ -175,8 +176,8 @@ def range_hash(key)
       end
 
       def field_hash(key)
-        val = query[key]
         attr, operation = key.to_s.split('.')
```
Comment (pboling) [inline lib/dynamoid/criteria/chain.rb]: We should change it here as well.

File: lib/dynamoid/criteria/chain.rb [hunk 2]
```diff
@@ -156,9 +156,10 @@ def records_via_scan
       end
 
       def range_hash(key)
-        val = query[key]
+        attr, operation = key.to_s.split('.')
```
Comment (andrykonchin) [inline lib/dynamoid/criteria/chain.rb]: @pboling please show me this in the Rails documentation. Haven't found it.

Comment (pboling) [inline lib/dynamoid/criteria/chain.rb]: @andrykonchin Sorry, I misread my documentation tool.  It is core Ruby, not Rails.  It is defined in the core Ruby `Module` object.

http://ruby-doc.org/core-2.4.1/Module.html#method-i-attr

Comment (andrykonchin) [inline lib/dynamoid/criteria/chain.rb]: 👍

**Comment by andrykonchin** [comment]
Done

**Pull request merged**